### PR TITLE
Nearley for Parsing

### LINKED ISSUE #1179: ID filter mapping slightly off
On Scryfall, `id:colors` translates to `id<=colors`, but we translate it to `id=colors`

### PR DESCRIPTION
**Pull request by ruler501**
Introduces a new parsing library to write the parsers in. This should enable much greater extensibility and hopefully make the code easier to understand. Next steps would be enabling a syntax and semantic aware autocomplete for cube filters and card filters. This PR ports the card and cube filters to use the nearley library and they actually end up sharing a good amount of the code. It also includes updated tests for both kinds of filters.

This does add a new build step to compile the grammars to javascript from their native format.

Depends on #1217 which depends on #1218.

Fixes #1179 and #1291.

### CODE REVIEW

File: __tests__/src/filtering/filterCards.test.js [hunk 1]
```diff
@@ -589,298 +457,196 @@ describe('filter', () => {
...
-      beforeEach(() => {
-        tokens = [];
-      });
-
       it('filters by card type', () => {
-        Filter.tokenizeInput('type=creature', tokens);
-        const creatureFilter = [Filter.parseTokens(tokens)];
-        const creatureCards = exampleCube.cards.filter((card) => Filter.filterCard(card, creatureFilter));
+        const { filter } = makeFilter('type:creature');
+        console.warn(filter);
```
Comment (RKelln) [inline __tests__/src/filtering/filterCards.test.js:626]: console.warn

**Comment by phulin** [comment]
I am open to this approach, but there are a variety of unrelated changes in this PR that make it very difficult to review. Can you separate this PR from the card properties one?

**Comment by ruler501** [comment]
@phulin it's very hard to separate them since the card properties pr is what gives us a consistent generic way to refer to properties of the cards.

**Comment by phulin** [comment]
I appreciate that, but I am still skeptical about the approach of that PR. If you want to keep coming back to this while that PR gets revised, be my guest - but we could merge this much sooner otherwise.

**Comment by ruler501** [comment]
@phulin I'll try and see what I can do without too much extra work.

File: src/pages/CubeListPage.js [hunk 2]
```diff
@@ -62,7 +61,8 @@ const CubeListPageRaw = ({
   }));
 
   const filteredCards = useMemo(() => {
-    return filter.length > 0 ? cube.cards.filter((card) => Filter.filterCard(card, filter)) : cube.cards;
+    console.warn(filter);
```
Comment (RKelln) [inline src/pages/CubeListPage.js:64]: console

File: nearley/cardFilters.ne [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+};
+%} # %}
+
+cmcCondition -> "cmc"i integerOpValue {% ([, valuePred]) => genericCondition('cmc', (card) => card.cmc ?? card.details.cmc, valuePred) %}
+
+colorCondition -> ("c"i | "color"i | "colors"i) colorCombinationOpValue {% ([, valuePred]) => genericCondition('colors', (card) => card.details.colors, valuePred) %}
+
+colorIdentityCondition -> ("ci"i | "id"i | "identity"i | "coloridentity" | "color_identity"i) colorIdentityOpValue {% ([, valuePred]) => genericCondition('color_identity', (card) => card.colors ?? card.details.color_identity, valuePred) %}
+
+typeCondition -> ("t"i |  "type"i | "type_line"i) stringContainOpValue {% ([, valuePred]) => genericCondition('type_line', (card) => card.type_line ?? card.details.type, valuePred) %}
```
Comment (RKelln) [inline nearley/cardFilters.ne:70]: I'd add `typeline` to this too (as matching `coloridentity`)

File: nearley/values.ne [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+    ("brown"i | "colorless"i) {% () => [] %}
+  | "white"i {% () => ['w'] %}
+  | "blue"i {% () => ['u'] %}
+  | "black"i {% () => ['b'] %}
+  | "red"i {% () => ['r'] %}
+  | "green"i {% () => ['g'] %}
+  | "azorious"i {% () => ['w', 'u'] %}
+  | "dimir"i {% () => ['u', 'b'] %}
+  | "rakdos"i {% () => ['b', 'r'] %}
+  | "gruul"i {% () => ['r', 'g'] %}
```
Comment (RKelln) [inline nearley/values.ne:105]: Add the misspelling of `grul` (if it doesn't break anything by adding it)?

Comment (RKelln) [reply nearley/values.ne:105]: Similarly `izet`

File: webpack.common.js [hunk 5]
```diff
@@ -73,7 +73,7 @@ const serverConfig = merge(config, {
     'pages/DashboardPage': './src/pages/DashboardPage.js',
     'utils/Card': './src/utils/Card.js',
     'utils/draftutil': './src/utils/draftutil.js',
-    'utils/Filter': './src/utils/Filter.js',
+    'filtering/filterCards': './src/filtering/filterCards.js',
```
Comment (phulin) [inline webpack.common.js:76]: Can you follow the capitalized naming convention for all of the files in `src` here?

File: routes/root.js [hunk 6]
```diff
@@ -185,7 +184,7 @@ router.get('/dashboard', async (req, res) => {
     return res.render('dashboard', {
       reactHTML:
         NODE_ENV === 'production'
-          ? await ReactDOMServer.renderToString(React.createElement(DashboardPage, reactProps))
+          ? ReactDOMServer.renderToString(React.createElement(DashboardPage, reactProps))
```
Comment (phulin) [inline routes/root.js:187]: This breaks server-side rendering. Please revert.

File: routes/root.js [hunk 7]
```diff
@@ -266,7 +265,7 @@ router.get('/dashboard/decks/:page', async (req, res) => {
   }
 });
 
-router.get('/landing', async (req, res) => {
+router.get('/landing', async (_, res) => {
```
Comment (phulin) [inline routes/root.js:268]: Can you remove all of these random changes?

File: routes/root.js (hunk 6 above)
Comment (ruler501) [reply routes/root.js:187]: My static analysis says this isn't an async function so await shouldn't do anything here.

File: routes/root.js (hunk 7 above)
Comment (ruler501) [reply routes/root.js:268]: Yeah. I just had my IDE autofix all warnings it found like unused params and unneeded awaits.

Comment (phulin) [reply routes/root.js:268]: Can you turn off that setting and remove these changes?

**Comment by dekkerglen** [comment]
Some CI failures, can you merge master in and get the Github actions checks going?

**Comment by dekkerglen** [comment]
Conflicts with package.json

**Pull request merged**